rules_daml/daml-build.bzl

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) 2020 The DAML Authors. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+def file_of_target(k):
+    [file] = k.files.to_list()
+    return file
+
+_damlc = attr.label(
+    allow_single_file = True,
+    default = Label("//compiler/damlc"),
+    executable = True,
+    cfg = "host",
+)
+
+def _daml_build_impl(ctx):
+    name = ctx.label.name
+    config = ctx.file.config
+    srcs = ctx.files.srcs
+    dd = ctx.attr.dar_dict
+    damlc = ctx.file._damlc
+    project_root = config.dirname
+    input_dars = [file_of_target(k) for k in dd.keys()]
+    output_dar = ctx.actions.declare_file(name + ".dar")
+    ctx.actions.run_shell(
+        tools = depset([damlc]),
+        inputs = depset([config] + srcs + input_dars),
+        outputs = [output_dar],
+        progress_message = "Building DAML project %s" % name,
+        command = """
+            set -eou pipefail
+            {moves}
+            {damlc} build --project-root {project_root} -o {dot_dot}/{output_dar}
+            file {output_dar}
+        """.format(
+            moves = "\n".join([
+                "mv {a} {b}".format(
+                    a = file_of_target(k).path,
+                    b = project_root + "/" + v,
+                )
```

Comment:
So `cp` should be ok?

Reply:
Yes, that's safe

Reply:
I changed the `mv` to `cp` and the windows build failed with 
```cp: cannot create regular file 'docs/source/upgrade/example/coinV2/coin-1.0.0.dar': Permission denied``` !!
What happened here?

Reply:
I didn't find the old Azure logs to confirm, but if this is running on the same node, then this could be a collision with the moved file from the previous run. Looking at the command
```
cp bazel-out/x64_windows-opt/bin/docs/daml-upgrade-example-v1.dar docs/sour...
```
this is copying a generated file into the source tree. I don't have a Windows machine in front of me right now to confirm, but Bazel usually makes generated files read-only. So, this is probably trying to overwrite a write protected file. We've had similar issues with the Cabal rules in the past, and had to call `cabal clean` in the end to work around it. See https://github.com/tweag/rules_haskell/pull/1156.

To avoid this issue this command should probably also set a `trap` to remove the copied files.

Reply:
I changed to use `cp -f`, but I wonder if to be properly resilient requires doing the daml build in a temporary directory. There was another build failure complaining about files in the package database under .daml - and the failure just went away on re-run.

Reply:
It’s been a while so I don’t remember the exact errors but looking at the git history we did something very similar for the genrules around triggers, see https://github.com/digital-asset/daml/pull/2816

Reply:
Thanks, I'll take a look at this tomorrow...

Reply:
> I changed to use `cp -f`, but I wonder if to be properly resilient requires doing the daml build in a temporary directory. There was another build failure complaining about files in the package database under .daml - and the failure just went away on re-run.

`cp -f` will avoid the permission denied issue, but it will still leave files behind in the source tree. That's an issue, because they might be picked up by a `glob` on the next run. If nothing else, it'll mess with the cache. Yeah, a temporary directory would make certain that we don't leave files behind.